File: natsu/src/lib/conversation.ts
import { invoke } from '@tauri-apps/api/core';

// Types matching Rust models
export type MessageRole = 'user' | 'assistant' | 'system';

export interface Message {
  id: string;
  conversation_id: string;
  role: MessageRole;
  content: string;
  created_at: number;
}

export interface Conversation {
  id: string;
  title: string;
  created_at: number;
  updated_at: number;
}

export interface ConversationWithMessages {
  id: string;
  title: string;
  created_at: number;
  updated_at: number;
  messages: Message[];
}

// API wrapper functions
export const conversationApi = {
  create: async (title: string): Promise<Conversation> => {
    return invoke('create_conversation', { title });
  },

  list: async (): Promise<Conversation[]> => {
    return invoke('list_conversations');
  },

  get: async (id: string): Promise<ConversationWithMessages> => {
    return invoke('get_conversation', { id });
  },

  addMessage: async (
    conversationId: string,
    role: MessageRole,
    content: string
  ): Promise<Message> => {
    return invoke('add_message', { conversationId, role, content });
  },

  delete: async (id: string): Promise<void> => {
    return invoke('delete_conversation', { id });
  },

  rename: async (id: string, title: string): Promise<Conversation> => {
    return invoke('rename_conversation', { id, title });
  },
};

File: natsu/src-tauri/src/models/conversation.rs
use serde::{Deserialize, Serialize};

/// Role of a message in a conversation
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum MessageRole {
    User,
    Assistant,
    System,
}

impl MessageRole {
    pub fn as_str(&self) -> &'static str {
        match self {
            MessageRole::User => "user",
            MessageRole::Assistant => "assistant",
            MessageRole::System => "system",
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "user" => Some(MessageRole::User),
            "assistant" => Some(MessageRole::Assistant),
            "system" => Some(MessageRole::System),
            _ => None,
        }
    }
}

/// A single message in a conversation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub id: String,
    pub conversation_id: String,
    pub role: MessageRole,
    pub content: String,
    pub created_at: i64,
}

/// A conversation (without messages, for listing)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Conversation {
    pub id: String,
    pub title: String,
    pub created_at: i64,
    pub updated_at: i64,
}

/// A conversation with all its messages
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationWithMessages {
    pub id: String,
    pub title: String,
    pub created_at: i64,
    pub updated_at: i64,
    pub messages: Vec<Message>,
}

/// Request to create a new conversation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateConversationRequest {
    pub title: String,
}

/// Request to add a message to a conversation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddMessageRequest {
    pub conversation_id: String,
    pub role: MessageRole,
    pub content: String,
}

/// Request to rename a conversation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RenameConversationRequest {
    pub id: String,
    pub title: String,
}
